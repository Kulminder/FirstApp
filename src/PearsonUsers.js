import React, { Component } from 'react';
import UserInfo from './UserInfo';

const fetchAllUserURL = 'https://reqres.in/api/users?page=1&per_page=10' //URL to fetch all Users

export class PearsonUsers extends Component {
  constructor(props) {
    super(props);

    this.state = {
      users: [
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        },
        {
          id: 5,
          first_name: "Charles",
          last_name: "Morris",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
        },
        {
          id: 6,
          first_name: "Tracey",
          last_name: "Ramos",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
        }],
      error: null
    };
  };

  componentDidMount() {
    fetch(fetchAllUserURL)
      .then(response => {
        if (response.ok) {return response.json();
        } else {throw new Error('Something went wrong ...');}
      })
      .then(data =>
        this.setState((prevState) => {
          return {
            users: [...prevState.users, ...data.data],
          }}, () => { this.getUniqueUsers() })
      )
      .catch(error => this.setState({ error }));

  }

  getUniqueUsers = () => {
    this.setState((prevState) => {
      return {
        users: prevState.users.map(e => e['id'])
          .map((e, i, final) => final.indexOf(e) === i && i)
          .filter(e => prevState.users[e]).map(e => prevState.users[e])
      }
    })
  }

  handleDeleteUser = (event, userID) => {
    const userList = this.state.users.filter(user => user.id !== userID);
    this.setState({ users: userList });
    event.preventDefault();
  }

  render() {
    const { users, error } = this.state;

    if (error) {
      return <p>{error.message}</p>;
    }

    let userList = users.map((user, index) =>
      <UserInfo user={user} key={index} onDelete={(e) => this.handleDeleteUser(e, user.id)} />
    )
    return (
      <div>
        <h1 className='userTitle'>Pearson User Management</h1>
        <ul className="persons">
          {userList}
        </ul>
      </div>
    );
  }
}
