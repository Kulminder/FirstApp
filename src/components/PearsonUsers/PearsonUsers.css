.person-list-area{
  background: #d8d8d8;
  padding:60px 20px 20px;
  overflow: hidden;
}