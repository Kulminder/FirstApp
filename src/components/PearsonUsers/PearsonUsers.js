import React, { Component } from 'react';
import './PearsonUsers.css';
import UserInfo from '../UserInfo/UserInfo';
import userDataService from '../../services/users-service'

export class PearsonUsers extends Component {
  constructor(props) {
    super(props);

    this.state = {
      users: [
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        },
        {
          id: 5,
          first_name: "Charles",
          last_name: "Morris",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
        },
        {
          id: 6,
          first_name: "Tracey",
          last_name: "Ramos",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
        }],
      error: null
    };
  };

  componentDidMount() {
    this.getAllUsers();
  }

  getAllUsers = () => {
    userDataService.fetchAllUsers()
      .then(response => {
        if (response.data) {
          return this.filterUniqueUsers(response.data)
        } else { throw new Error('Something went wrong ...'); }
      })
      .catch(error => this.setState({ error }));
  }

  filterUniqueUsers = (response) => {
    this.setState((prevState) => {
      let users = [...prevState.users, ...response.data];
      return {
        users: users.map(e => e['id'])
          .map((e, i, final) => final.indexOf(e) === i && i)
          .filter(e => users[e]).map(e => users[e])
      }
    })
  }

  handleDeleteUser = (event, userIndex) => {
    const userList = [...this.state.users]
    userList.splice(userIndex, 1);
    this.setState({ users: userList });
    event.preventDefault();
  }

  render() {
    if (this.state.error) {
      return <p>{this.state.error.message}</p>;
    }
    let users = null;
    if (this.state.users && this.state.users.length) {
      users = (
        <UserInfo
          onDelete={this.handleDeleteUser}
          users={this.state.users} />
      )
    }

    return (
      <React.Fragment>
        <h1 className='page-title'>Pearson User Management</h1>
        <div className="person-list-area">
             {users}
        </div>
      </React.Fragment>
    );
  }
}
