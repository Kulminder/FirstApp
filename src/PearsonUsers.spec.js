import React from "react";
import { shallow } from 'enzyme';
import { PearsonUsers } from "./PearsonUsers";
import toJson from 'enzyme-to-json';
import sinon from 'sinon';

const setup = () => shallow(<PearsonUsers />);

describe("PearsonUsers component", () => {

  beforeEach(() => {
    sinon.stub(console, 'error');
  });

  afterEach(() => {
    console.error.restore();
  });

  it('instance should not be null', () => {
    const wrapper = setup();
    const instance = wrapper.instance();
    expect(instance).not.toBe(null);
  });

  it('should renders correctly', () => {
    const wrapper = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should renders a h1', () => {
    const wrapper = setup();
    expect(wrapper.find('h1').text()).toEqual("Pearson User Management");

  });

  it('should have User Info as child component', () => {
    const wrapper = setup();
    expect(wrapper.find('UserInfo')).toBeDefined();
  });

  it('should have User Info as child component', () => {
    const wrapper = setup();
    expect(wrapper.find('UserInfo')).toBeDefined();
  });

  it("handleDeleteUser function should be defined", () => {
    const wrapper = setup();
    const event = { preventDefault: () => { } };
    expect(wrapper.instance().handleDeleteUser(event, 4)).toBeUndefined();
  });

  // it('should update state', (done) => {
  //   const wrapper = mount(<PearsonUsers />);

  //   const initialUserList = [
  //     {
  //       id: 4,
  //       first_name: "Eve",
  //       last_name: "Holt",
  //       avatar:
  //         "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
  //     },
  //   ]

  //   const result = {
  //     data: {
  //       users: {
  //         id: 5,
  //         first_name: "George",
  //         last_name: "Bush",
  //         avatar:
  //           "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
  //       },
  //     }
  //   }
  //   const promise = Promise.resolve(result);

  //   expect(wrapper.state().users).toEqual(initialUserList);

  //   promise.then(() => {
  //     wrapper.update();
  //     console.log('wrapper.state().users', wrapper.state().users)
  //     //expect(wrapper.state().users).toEqual(props.data.data);

  //     done();
  //   });
  // });
});