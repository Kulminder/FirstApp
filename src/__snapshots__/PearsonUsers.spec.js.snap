// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PearsonUsers component should renders correctly 1`] = `
<div>
  <h1
    className="userTitle"
  >
    Pearson User Management
  </h1>
  <ul
    className="persons"
  >
    <UserInfo
      key="0"
      onDelete={[Function]}
      user={
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
          "first_name": "Eve",
          "id": 4,
          "last_name": "Holt",
        }
      }
    />
    <UserInfo
      key="1"
      onDelete={[Function]}
      user={
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
          "first_name": "Charles",
          "id": 5,
          "last_name": "Morris",
        }
      }
    />
    <UserInfo
      key="2"
      onDelete={[Function]}
      user={
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
          "first_name": "Tracey",
          "id": 6,
          "last_name": "Ramos",
        }
      }
    />
  </ul>
</div>
`;
